Improvements to STM text (#6398)
quarkusio/quarkus: docs/src/main/asciidoc/software-transactional-memory.adoc

one or more of the threads may be forced to abort if there have been conflicting updates.

* STM objects have state but it doesn't need to be persistent (durable). In fact the
default behaviour is for objects managed within transactional memory to be volatile, such that
if the service or microservice within which they are being used crashes or is spawned elsewhere, e.g.,
by a scheduler, all state in memory is lost and the objects start from scratch. But surely you get this and more
with JTA (and a suitable transactional datastore) and don't need to worry about restarting your application?
Not quite. There's a trade-off here: we're doing away
with persistent state and the overhead of reading from and then writing (and sync-ing) to the datastore during each
transaction. This makes updates to (volatile) state very fast but you still get the benefits of atomic updates
across multiple STM objects (e.g., objects your team wrote then calling objects you inherited from another team and requiring
them to make all-or-nothing updates), as well as consistency
and isolation in the presence of concurrent threads/users (common in distributed microservices architectures).

> Furthermore, not all stateful applications need to be durable - even when JTA transactions are used, the tend to be the

Reviewer: ```suggestion
Furthermore, not all stateful applications need to be durable - even when JTA transactions are used, it tends to be the
```
but it might just be me not being native.

Author: Good catch.

</dependencies>
--

== Defining STM-aware classes

In order for the STM subsytem to have knowledge about which classes are to be managed within the context
of transactional memory it is necessary to provide a minimal level of instrumentation. This occurs by

> categorising STM-aware and STM-unware classes through an interface bounday; specifically all STM-aware objects

Reviewer: ```suggestion
categorising STM-aware and STM-unaware classes through an interface boundary; specifically all STM-aware objects
```

}
--

Classes which implement this interface are able to use additional annotations from Narayana to tell the STM
subsystem about things such as whether a method will modify the state of the object, or what state variables
within the class should be managed transactionally, e.g., some instance variables may not need to be rolled back

> if a transaction aborts.. As mentioned earlier, if those annotations are not present then defaults are chosen to

Reviewer: ```suggestion
if a transaction aborts. As mentioned earlier, if those annotations are not present then defaults are chosen to
```

guarantee safety, such as assuming all methods will modify state.

[source,java]
--
public class FlightServiceImpl implements FlightService {
    @ReadLock
    public int getNumberOfBookings() { ... }
    public void makeBooking(String details) {...}
    
    @NotState
    private int timesCalled;
}
--

> For example, by using the ReadLock annotation on the `getNumberOfBookings` method, we are able to tell the

Reviewer: ```suggestion
For example, by using the `@ReadLock` annotation on the `getNumberOfBookings` method, we are able to tell the
```

STM subsystem that no state modifications will occur in this object when it is used in the transactional

> memory. Also, the NotState annotation tells the system to ignore `timesCalled` when transactions commit or

Reviewer: ```suggestion
memory. Also, the `@NotState` annotation tells the system to ignore `timesCalled` when transactions commit or
```

abort, so this value only changes due to application code.

Please refer to the Narayana guide for details of how to exert finer grained control over the transactional
behaviour of objects that implement interfaces marked with the `@Transactional` annotation.

== Creating STM objects

The STM subsystem needs to be told about which objects it should be managing. The Quarkus (aka Narayana) STM implementation
does this by providing containers of transactional memory within which these object instances reside. Until an object
is placed within one of these STM containers it cannot be managed within transactions and any state changes will
not possess the A, C, I (or even D) properties.

Note, the term "container" was defined within the STM implementation years before Linux containers came along. It may

> be confusing to use especially in a Kuberenetes native environment such as Quarkus, but hopefully

Reviewer: ```suggestion
be confusing to use especially in a Kubernetes native environment such as Quarkus, but hopefully
```

<1> You need to tell each Container about the type of objects for which it will be responsible. In this example
    it will be instances that implement the FlightService interface.

> <2> Then you create an instance that implements FlightService. You should not use it directly at this stage because

Reviewer: ```suggestion
<2> Then you create an instance that implements `FlightService`. You should not use it directly at this stage because
```

Author: Must've been there originally because I haven't edited that text yet ...

    access to it is not being managed by the STM subsystem.
<3> To obtain a managed instance, pass the original object to the STM `container` which then returns a reference
    through which you will be able perform transactional operations. This reference can be used safely from multiple threads.

== Defining transaction boundaries

Once an object is placed within an STM container the application developer can manage the scope of transactions
within which it is used. There are some annotations which can be applied to the STM-aware class to have the
container automatically create a transaction whenever a specific method is invoked.

=== Declarative approach

If the `@NestedTopLevel` or `@Nested` annotation is placed on a method signature then the STM container will
start a new transaction when that method is invoked and attempt to commit it when the method returns. If there is

> a transaction already associated with the caling thread then each of these annotations behaves slightly differently:

Reviewer: ```suggestion
a transaction already associated with the calling thread then each of these annotations behaves slightly differently:
```
